### ISSUE TITLE
[Certificate Management] Inhibit self-signed cert generation & secret creation during cluster bootstrap

### ISSUE THREAD

**Issue by BarthV** [kubernetes-sigs/cluster-api#8255]
**User Story**

As a CAPI cluster admin & operator, I would like to easily provide custom certificates for my CAPI managed clusters **and** prevent CAPI controllers to generate any self signed certificates. (So I'd also like to be sure that no race condition occurs during the time kube secrets are missing)

**Detailed Description**

Cluster API book says that [we can perfectly provide custom CA as secrets.](https://cluster-api.sigs.k8s.io/tasks/bootstrap/kubeadm-bootstrap.html?highlight=certificate#certificate-management) . But in fact, there's nearly 0 examples or guides to explain how to proceed.

So we started to dig cluster-api sources to understand how certificates are managed by Control Plane & Kubeadm Controller.

We discovered ([here](https://github.com/kubernetes-sigs/cluster-api/blob/main/util/secret/certificates.go#L264-L278)) that, during CAPI cluster reconciliation, there's no way to prevent Cluster-API to generate certificate and store them in kube secrets. If I ever want to manage these secrets by myself without an additional workflow (e.g. apply secrets, wait, then apply CAPI cluster specs), I'd expect an option to tell CAPI not to handle self-signed cert generation & fails its reconciliation untils secrets are available.

TLDR; I'd expect some kind a new flag in specs to inhibit self-signed certificate generation done by CAPI controllers. That would enable my teams to manage clusters certificates without having to handle possible race conditions on our side.

/kind feature

**Comment by killianmuldoon** [comment]
Maybe I'm mistaken, but as I understand the model here if you provider own certs they will not be generated in this part of the code. So if you create your certs when creating your Cluster they shouldn't be created at this point. Is there some issue with the ordering in your use case?

It's definitely a good idea to add some documentation on how to bring your own certs to CAPI though.

**Comment by BarthV** [comment]
> Maybe I'm mistaken, but as I understand the model here if you provider own certs they will not be generated in this part of the code. So if you create your certs when creating your Cluster they shouldn't be created at this point. Is there some issue with the ordering in your use case?
> 
> It's definitely a good idea to add some documentation on how to bring your own certs to CAPI though.

You're definitely right if your working with staticly provisionned secrets. In this case, you can apply your static secrets then capi cluster, and it works well.

But I'd like to manage this secrets using a secret manager or using cert-manager signed certs. In this case I cannot guarantee when the secret will be available. So I'll definitely have to care about race conditions (and love to be able to prevent CAPI to ever generate them if missing)

**Comment by killianmuldoon** [comment]
Understood. I think one of the complexities with this is that currently the secrets can either be generated by KCP (Kubeadm Control Plane) or by Kubeadm Bootstrap  (CAPBK). 

If KCP doesn't generate the secrets they will be generated by CAPBK. The problem here will be in syncing the behaviour across the two providers IMO - an identical flag will need to be passed to both to ensure they don't generate secrets.

I think a flag would be a welcome idea though - and we could consider leveraging it to control Kubeadm Bootstrap's cert generation.

/triage accepted

**Comment by fabriziopandini** [comment]
> But I'd like to manage this secrets using a secret manager or using cert-manager signed certs. In this case I cannot guarantee when the secret will be available. So I'll definitely have to care about race conditions (and love to be able to prevent CAPI to ever generate them if missing)

I have some concerns about bringing into CAPI considerations about race conditions due to external systems that can be used to generate secrets, which is something out of the scope of CAPI, it ultimately seems we are solving the issue in the wrong layer.

Also, adding on top of @killianmuldoon comment, let's keep in the radar that we have to consider providing and maintaining a new E2E test for the new scenario - and this will require us to similate an external system providing certificates -, and potentially improve unit test coverage on some parts of CAPBK and KCP before adding more complexity on top of it.

**Comment by hrak** [comment]
#9168 duplicate?

**Comment by killianmuldoon** [comment]
Looks like it - thanks for pointing this out.

**Comment by fabriziopandini** [comment]
For reference, [my comment](https://github.com/kubernetes-sigs/cluster-api/issues/9168#issuecomment-1676302605) on the duplicate PR mirroring what I wrote above

**Comment by fabriziopandini** [comment]
The Cluster API project currently lacks enough active contributors to adequately respond to all issues and PRs.

No activity since 1Yr now, no interest from other community members/any contributor
/close
